### PR TITLE
Single consolidated executable

### LINKED ISSUE #2336: Single consolidated executable
This is something I'd like to do before 1.0:

``` text
> psc build 'src/**/*.purs'
> psc bundle 'output/*/*.js'
> psc docs
> psc publish
> psc repl
> psc ide
```

etc.

It should also reduce the size of the binary distribution a bit.

### PR DESCRIPTION
**Pull request by paf31**
Fixes #2336. This also removes `psc-ide-client`, which had been discussed previously.

### CODE REVIEW

**Comment by kRITZCREEK** [comment]
Looking good! I've pr'd #2634 against this branch to fix the test failures. We said we wanted to provide wrappers around the consolidated executable to smoothen the breakage or do we just go full burning bridges with 0.11?

File: src/Language/PureScript/Options.hs [hunk 1]
```diff
@@ -1,42 +1,20 @@
...
+  { optionsVerboseErrors :: Bool
+  -- ^ Verbose error message
   , optionsNoComments :: Bool
-    -- |
-    -- Generate soure maps
+  -- ^ Remove the comments from the generated js
   , optionsSourceMaps :: Bool
-    -- |
-    -- Dump CoreFn
+  -- ^ Generate soure maps
```
Comment (hdgarrood) [inline src/Language/PureScript/Options.hs]: Typo?

**Comment by hdgarrood** [comment]
Perhaps we could provide shell scripts named after the old executables for backwards compatibility in the binary bundles, just for the 0.11 series? They'd only need to be a few lines I think.

For true backwards compat we would also need to have `psc` work like the new `psc build`, I guess that could be triggered if the first argument doesn't match any of the sub-commands.

File: app/Command/Docs.hs [hunk 2]
```diff
@@ -218,20 +215,15 @@ buildOptions (fmt, input, mapping) =
...
-  version = abortOption (InfoMsg (showVersion Paths.version)) $ long "version" <> help "Show the version number" <> hidden
+command :: Opts.Parser (IO ())
+command = (buildOptions >=> docgen) <$> (version <*> Opts.helper <*> pscDocsOptions) where
+  version :: Opts.Parser (a -> a)
+  version = Opts.abortOption (Opts.InfoMsg (showVersion Paths.version)) $ Opts.long "version" <> Opts.help "Show the version number" <> Opts.hidden
+
+infoModList :: Opts.InfoMod a
+infoModList = Opts.fullDesc <> footerInfo where
+  footerInfo = Opts.footerDoc $ Just $ PP.vcat
+                 [ examples, PP.empty, PP.text ("psc-docs " ++ showVersion Paths.version) ]
```
Comment (hdgarrood) [inline app/Command/Docs.hs]: Oh wait, should we remove this reference to `psc-docs`?

File: app/Command/Compile.hs [hunk 3]
```diff
@@ -89,94 +84,66 @@ globWarningOnMisses warn = concatMapM globWithWarning
...
-  where
-  opts        = info (version <*> helper <*> pscMakeOptions) infoModList
-  infoModList = fullDesc <> headerInfo <> footerInfo
-  headerInfo  = header   "psc - Compiles PureScript to Javascript"
-  footerInfo  = footer $ "psc " ++ showVersion Paths.version
-
-  version :: Parser (a -> a)
-  version = abortOption (InfoMsg (showVersion Paths.version)) $ long "version" <> help "Show the version number" <> hidden
+command :: Opts.Parser (IO ())
+command = compile <$> (version <*> Opts.helper <*> pscMakeOptions) where
```
Comment (hdgarrood) [inline app/Command/Compile.hs]: I'm not sure the sub commands need to be able to respond to `--version`. Having `psc --version` work is enough IMO, we don't also need to support `psc build --version` etc.

File: src/Language/PureScript/Options.hs (hunk 1 above)
Comment (paf31) [inline src/Language/PureScript/Options.hs]: Yes, thanks.

File: app/Command/Compile.hs (hunk 3 above)
Comment (paf31) [inline app/Command/Compile.hs]: True :)

File: app/Command/Docs.hs (hunk 2 above)
Comment (paf31) [inline app/Command/Docs.hs]: Yes, looks like it.

**Comment by paf31** [comment]
I'll fix the remaining bits on the other branch, after merging this. I also need to update the bundle script.

**Comment by kRITZCREEK** [comment]
I just checked and it seems like `psc-ide-vim` still relies on `psc-ide-client`. Maybe it would be best to keep it around behind `purs ide --client`.

**Comment by paf31** [comment]
@kRITZCREEK Sounds good, we can add it back.

**Pull request merged**